### PR TITLE
feat: batch-aware expert pruning (XShare) for MoE models

### PR DESCRIPTION
**Pull request by hai-meh-cs**
## Summary

Implements batch-aware expert pruning ([XShare](https://arxiv.org/abs/2602.07265)) for MoE models. Before top-k routing, aggregates expert demand across the batch and masks low-demand experts so the fused MoE kernel skips loading their weights from HBM.

- Adds `--moe-config` CLI flag with `MoEConfig` / `MoEPruningConfig` dataclasses
- Fused Triton kernel for global expert selection + per-row topk masking
- Config persistence through vLLM's `set_current_vllm_config` context manager
- **Decode-only gating** (`max_batch=48`): pruning is skipped during prefill to preserve KV cache quality

RFC: https://github.com/vllm-project/vllm/issues/35550

### 128-expert model (1x H200, gpt-oss-120b)

Setup: `vllm bench serve`, 200 prompts, random in=128 out=256, rate=inf, `expert_budget=24`

| Concurrency | Baseline (tok/s) | XShare (tok/s) | Throughput | TPOT Reduction |
|:-:|:-:|:-:|:-:|:-:|
| 4 | 530 | 559 | +5.4% | -5.2% |
| 8 | 866 | 1,006 | +16.2% | -13.9% |
| 16 | 1,342 | 1,835 | +36.8% | -26.9% |
| 32 | 1,996 | 3,054 | +53.0% | -34.6% |

### 32-expert model (1x H200, gpt-oss-20b)

Setup: `expert_budget=12`, 64 prompts, random in=128 out=256, rate=inf

| Concurrency | Baseline (tok/s) | XShare (tok/s) | Throughput | TPOT Reduction |
|:-:|:-:|:-:|:-:|:-:|
| 4 | 869 | 857 | -1.5% | +1.8% |
| 8 | 1,503 | 1,660 | +10.4% | -9.5% |
| 16 | 2,719 | 3,122 | +14.8% | -13.5% |
| 32 | 4,464 | 5,273 | +18.1% | -15.8% |

### Rate-limited online serving (1x H200, gpt-oss-120b, rate=8 req/s)

| Workload | Baseline (tok/s) | XShare budget=24 (tok/s) | Improvement |
|---|:-:|:-:|:-:|
| Short (in=128, out=128) | 780 | 824 | +5.7% |
| Medium (in=512, out=256) | 1,162 | 1,321 | +13.6% |
| Long (in=1024, out=512) | 1,508 | 1,850 | +22.7% |

## Quality validation (lm-eval-harness)

Setup: 8x H200 (p5e.48xlarge), TP=8, max-model-len=32768, max-num-seqs=64, `num_concurrent=50`.
All pruning configs use decode-only gating (`max_batch=48`).

**MMLU-Pro (5-shot):**

| Config | expert_budget | top_per_token (k0) | Score | Delta vs baseline |
|---|:-:|:-:|:-:|:-:|
| Baseline (no pruning) | n/a | n/a | 0.5132 | n/a |
| budget=24, k0=1 | 24 | 1 | 0.5157 | +0.25% |
| budget=12, k0=1 | 12 | 1 | 0.5147 | +0.15% |
| budget=12, k0=2 | 12 | 2 | 0.5147 | +0.15% |
| budget=32, k0=1 | 32 | 1 | 0.5150 | +0.18% |

**GSM-8K (5-shot, exact_match strict):**

| Config | expert_budget | top_per_token (k0) | Score | Delta vs baseline |
|---|:-:|:-:|:-:|:-:|
| Baseline (no pruning) | n/a | n/a | 0.2699 | n/a |
| budget=24, k0=1 | 24 | 1 | 0.2813 | +1.14% |
| budget=12, k0=1 | 12 | 1 | 0.2745 | +0.46% |
| budget=12, k0=2 | 12 | 2 | 0.2782 | +0.83% |
| budget=32, k0=1 | 32 | 1 | 0.2623 | -0.76% |

All configs within noise of baseline. **Caveat:** with `num_concurrent=50` and `max_batch=48`, pruning fires intermittently during eval (decode batch sizes often approach or exceed 48). These results confirm pruning does not degrade quality when active, but do not stress-test sustained pruning at high intensity. Lower-concurrency eval or phase-based gating would provide stronger validation.

**Key finding:** `max_batch` gating is essential. Without it, pruning during prefill causes catastrophic quality loss (MMLU-Pro 51.3% to 26.9%). Prefill pruning corrupts the KV cache by routing prompt tokens through an incomplete expert set.

## Decode-only pruning: why and how

Pruning must be restricted to the decode phase. During prefill, all prompt tokens build the KV cache n/a routing them through a pruned expert set produces corrupted representations that cascade through the sequence.

Current gating: `max_batch=48` heuristic. During prefill M = prompt length (>> 48), so pruning is skipped. During decode M = concurrent sequences (typically ≤ 48), so pruning fires. This is empirical n/a phase-based gating via `max_query_len` in `CommonAttentionMetadata` is a planned improvement.

## Configuration

```python
@config
class MoEPruningConfig:
    enable: bool = False
    expert_budget: int | None = None  # absolute budget (overrides alpha)
    budget_alpha: float = 0.5         # fraction of experts to keep
    top_per_token: int = 0            # per-token top-k pre-selection (k0)
    group_budget: int = 0             # per-group expert budget (future)
    group_size: int = 0               # tokens per group (future)
    min_batch: int = 0                # skip pruning below this batch size
    max_batch: int = 48               # skip pruning above this (decode-only gating)
```

```bash
# Enable with absolute budget
vllm serve model --moe-config '{"pruning": {"enable": true, "expert_budget": 24}}'

# Enable with fractional budget
vllm serve model --moe-config '{"pruning": {"enable": true, "budget_alpha": 0.5}}'
```

## How to reproduce

```bash
# Baseline server
vllm serve openai/gpt-oss-120b \
    --tensor-parallel-size 1 --max-model-len 4096 --max-num-seqs 64 --port 8000

# XShare server
vllm serve openai/gpt-oss-120b \
    --tensor-parallel-size 1 --max-model-len 4096 --max-num-seqs 64 \
    --moe-config '{"pruning": {"enable": true, "expert_budget": 24}}' --port 8000

# Benchmark
vllm bench serve --backend openai --model openai/gpt-oss-120b \
    --dataset-name random --random-input-len 128 --random-output-len 256 \
    --num-prompts 200 --request-rate inf --max-concurrency 16

# Quality eval (TP=8 for speed)
vllm serve openai/gpt-oss-120b \
    --tensor-parallel-size 8 --max-model-len 32768 --max-num-seqs 64 \
    --moe-config '{"pruning": {"enable": true, "expert_budget": 24, "top_per_token": 1}}' \
    --port 8000

lm_eval --model local-completions \
    --model_args "model=openai/gpt-oss-120b,base_url=http://localhost:8000/v1/completions,num_concurrent=50,tokenized_requests=False" \
    --tasks mmlu_pro --num_fewshot 5 --output_path ./eval_results/mmlu_pro
```

## Test plan

- [x] Unit tests (`tests/test_xshare_pruning.py`)
- [x] Serving benchmarks on H200 (gpt-oss-120b 128E, gpt-oss-20b 32E)
- [x] CUDA graph compatibility verified
- [x] lm-eval quality validation (MMLU-Pro, GSM-8K n/a all configs within noise of baseline)
- [x] Tensor parallelism verified (TP=8, router replicated, identical pruning across ranks)
- [ ] Expert parallelism testing
- [ ] Lower-concurrency quality eval (ensure sustained pruning)
- [ ] Phase-based gating (replace max_batch heuristic)

Paper: [XShare: Batch-Aware Expert Selection for MoE Inference](https://arxiv.org/abs/2602.07265)

### CODE REVIEW

**Comment by stefanskiasan** [comment]
### Data point: no benefit on a host-bound MoE path (288 experts, ROCm/AITER)

I ported the core idea of this PR to a different setup to see whether it generalises, and it does not — which I think is a useful boundary condition for the RFC rather than a criticism of the work.

**Setup:** 8x MI350X (gfx950), ROCm 7.2.3, TP8, GLM-5.3-Flash-FP8 — **288 experts, top-8**, AITER fused MoE (`a8w8_blockscale`), not the `gpt_oss_triton_kernels_moe` path this PR touches. I implemented only the global-budget case (`global_sums = logits.sum(0)` → top-`ml` → `masked_fill`), applied to the router logits between the gate and the fused-MoE call, gated to decode-sized batches only.

Measured with `vllm bench serve`, realistic-prose dataset, three warmup passes (±5 % run-to-run):

| expert budget (of 288) | 32 concurrent | 96 concurrent |
|---|---|---|
| off | 1 748 tok/s | 2 890 tok/s |
| 144 (50 %) | 1 626 | 2 817 |
| 96 (33 %) | 1 745 | 2 969 |
| 64 (22 %) | 1 698 | 2 828 |

Everything is inside the noise band; no monotonic trend in either direction.

**Why it likely does not transfer.** On this configuration the MoE decode path is not HBM-bound. Measured at 96 concurrency: `rocm-smi` reports 100 % GPU utilisation and 95 % VRAM, but back-computing from step time gives roughly **1 % of peak FLOPs and ~7 % of peak bandwidth** (545 GB/s of 8 TB/s per GPU). A step takes 74 ms for 480 tokens where a bandwidth-bound step would take ~5 ms. The time goes into kernel launch and routing overhead — consistent with the ~9 300 kernel launches per output token reported for MoE decode versus 848 for dense.

If the bottleneck is launch/routing rather than expert weight traffic, skipping weight loads buys nothing, and the extra `topk` + `scatter` + `masked_fill` over an `[M, 288]` tensor is pure added work. The gains reported here on gpt-oss-120b/H200 are presumably on a path where weight traffic *is* the limit.

That suggests a useful gate for the feature: it should pay off in proportion to how bandwidth-bound the MoE step actually is. A cheap runtime signal (or simply documenting "expect gains only when the fused-MoE step is HBM-bound") would stop people like me from expecting it everywhere.

Independent evidence from the same box, in case it is useful: on this model tuning the fMoE kernel-selection table for the actually-occurring token bucket sizes gave **+27 %** at high concurrency, while every configuration-level flag we tried moved nothing. That is the signature of a launch-bound rather than bandwidth-bound regime.
